Added rate limiter to quarterdeck

pdamodaran opened the pull request:
### Scope of changes

Added IP based rate limiting middleware to quarterdeck to control the number of requests that can be submitted by a single IP address.  

There is an issue with a test that submits multiple requests that should be returning a 429 error but is returning 200 instead.

Feel free to suggest changes to the naming or other design patterns that don't follow convention.

Fixes SC-13314

### Type of change

- [x] new feature
- [ ] bug fix
- [ ] documentation
- [ ] testing
- [ ] technical debt
- [ ] other (describe)

### Acceptance criteria

Describe how reviewers can test this change to be sure that it works correctly. Add a checklist if possible.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [ ]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [ ]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.
- [ ] Are there any TODOs in this PR that should be turned into stories?

pdamodaran commented on pkg/quarterdeck/middleware/ratelimiter.go:
```diff
@@ -0,0 +1,117 @@
+	var onlyOnce sync.Once
+
+	var limiter *rate.Limiter
+	onlyOnce.Do(func() {
```
@bbengfort - I think this takes care of the double checked locking but let me know if this is not correct.

bbengfort replied:
It's possible that I'm misinterpreting this but it seems that the `GetLimiter` function grabs an Rlock, then unlock and calls AddIP; the `AddIP` function creates `onlyOnce` inside of the function; so that mutex is locally scoped. Inside of the function, `onlyOnce` will only be executed once, however multiple concurrent calls to `AddIP` will each have their own `onlyOnce` and will not be synchronized; so it will be overwrritten if the calls happen in quick succession (which may be what is happening in the tests). 

Double checked locking would be:

```
func (i *IPRateLimiter) AddIP(ip string) (*rate.Limiter) {
	i.mu.Lock()
	defer i.mu.Unlock()

	// Here is the second check, e.g. the double check
	if ipInfo, exists := i.ips[ip]; exists {
		return ipInfo.limiter
	}

	// Otherwise the condition from the RLock is still true, so create the limiter 
	limiter := rate.NewLimiter(i.limit, i.burst)
	i.ips[ip] = &ipInfo{limiter, time.Now()}
	return limiter 
}
```

pdamodaran replied:
understood - so that point behind the double check is to do the check in the `AddIP` method to ensure that ipInfo does not exist before adding the IP address to the map.

pdamodaran commented on pkg/quarterdeck/middleware/ratelimiter_test.go:
```diff
@@ -0,0 +1,91 @@
+				return
+			case <-ticker.C:
+				router.ServeHTTP(w, req)
+				//fmt.Println(w.Code)
```
I have print statements here and in the rate limiter code.  If you uncomment these and run this test, you will see that the test hits the section in the rate limiter code that sends the 429 error response but the code that actually gets written to the writer is 200.

bbengfort commented on pkg/quarterdeck/config/config.go:
```diff
@@ -32,6 +32,7 @@ type Config struct {
 	Database      DatabaseConfig
 	Token         TokenConfig
 	Sentry        sentry.Config
+	RateLimit     RatelimitConfig
```
```suggestion
	RateLimit     RatelimitConfig `split_words:"true"`
```

☝🏽 will change the environment variable from `$QUARTERDECK_RATELIMIT_BURST` to `$QUARTERDECK_RATE_LIMIT_BURST`.

bbengfort commented on pkg/quarterdeck/config/config.go:
```diff
@@ -50,6 +51,16 @@ type TokenConfig struct {
+// Burst: maximum number of tokens/requests in a "token bucket" and is initially full
+// Ttl: //number of minutes before an IP is removed from the ratelimiter map
+type RatelimitConfig struct {
+	Limit float64       `default0:"10"`
+	Burst int           `default0:"120"`
+	Ttl   time.Duration `default0:"5"`
```
This might make configuration a little easier:

```suggestion
	PerSecond float64       `default:"20" split_words:"true"`
	Burst int           `default:"120"`
	TTL   time.Duration `default:"5"`
```

Would you also mind updating the tests with the new config variables?

bbengfort commented on pkg/quarterdeck/middleware/ratelimiter.go:
```diff
@@ -0,0 +1,117 @@
+	i.mu.RUnlock()
+
+	// update the last seen time associated with the IP address
+	ipInfo.lastSeen = time.Now()
```
A concurrency guard on last seen might be required; e.g. a mutex in ipInfo.

bbengfort commented on pkg/quarterdeck/middleware/ratelimiter.go:
```diff
@@ -0,0 +1,117 @@
+				deleteList = append(deleteList, ip)
+			}
+		}
+		var mu sync.RWMutex
+		for _, ip := range deleteList {
+			mu.Lock()
+			delete(i.ips, ip)
+			mu.Unlock()
+		}
```
Make sure you're using the shared mutex!

```suggestion
		for _, ip := range deleteList {
			i.mu.Lock()
			delete(i.ips, ip)
			i.mu.Unlock()
		}
```

bbengfort commented on pkg/quarterdeck/middleware/ratelimiter.go:
```diff
@@ -0,0 +1,117 @@
+			c.Writer.Header().Add("Ratelimit-Request-Remote-Addr", c.Request.RemoteAddr)
+			c.Writer.Header().Add("Ratelimit-Remaining", fmt.Sprintf("%.2f", limiter.Tokens()))
+			//fmt.Println(http.StatusTooManyRequests)
+			c.AbortWithStatusJSON(http.StatusTooManyRequests, api.ErrorResponse(ErrRateLimit))
```
The abort should cancel all processing so I'm not 100% sure what is going on with the tests. I'll have to play around with it a little bit.

pdamodaran commented on pkg/quarterdeck/middleware/ratelimiter.go:
```diff
@@ -0,0 +1,117 @@
+		limiter := limiter.GetLimiter(c.Request.RemoteAddr)
+		if !limiter.Allow() {
+			c.Writer.Header().Add("Ratelimit-Request-Remote-Addr", c.Request.RemoteAddr)
+			c.Writer.Header().Add("Ratelimit-Remaining", fmt.Sprintf("%.2f", limiter.Tokens()))
```
I have made the suggested changes - let me know if you would like for me to add `X-RateLimit-Request-Remote-Addr` as well.

pdamodaran commented on pkg/quarterdeck/config/config.go:
```diff
@@ -50,6 +51,17 @@ type TokenConfig struct {
+// Burst: maximum number of tokens/requests in a "token bucket" and is initially full
+// NOTE: if Burst is not included the config, then ALL REQUESTS WILL BE REJECTED!
+// Ttl: //number of minutes before an IP is removed from the ratelimiter map
+type RatelimitConfig struct {
```
@bbengfort - if we don't include a value for `Burst` in the config, it defaults to 0 which means that every single request will get rejected.  Our tests will catch this, but should we set the defaults in `rate_limiter.go` if the values are not set?

pdamodaran commented:
@bbengfort - I saw the following error in the go 1.20 tests:

<img width="905" alt="Screenshot 2023-03-21 at 7 57 55 AM" src="https://user-images.githubusercontent.com/7106134/226610511-0e30c171-d0dc-49a0-9cd9-4750e09a7cf9.png">

I added an assertion that after the third request, the rate limit remaining would equal 0.00 but in this case the rate limit remaining was 0.01 which is possible based on the calculations done in the https://pkg.go.dev/golang.org/x/time@v0.3.0/rate code.

I removed the assertion to see if the next test will fail if we run into a scenario where the rate limit remaining is 0.01 but I have run the test a couple of times and it has passed each time.  

In fact, before I made that change, I ran the failed test a second time, and it succeeded.

bbengfort commented:
Yeh, timing in tests is hard with Go; we have a couple of tests that fail occasionally in the code because of timing issues. Usually we just try to make things flexible; e.g. instead of an == 0 assertion, maybe a <= 0.1 assertion; etc. 

> @bbengfort - I saw the following error in the go 1.20 tests:
> 
> <img alt="Screenshot 2023-03-21 at 7 57 55 AM" width="905" src="https://user-images.githubusercontent.com/7106134/226610511-0e30c171-d0dc-49a0-9cd9-4750e09a7cf9.png">
> 
> I added an assertion that after the third request, the rate limit remaining would equal 0.00 but in this case the rate limit remaining was 0.01 which is possible based on the calculations done in the https://pkg.go.dev/golang.org/x/time@v0.3.0/rate code.
> 
> I removed the assertion to see if the next test will fail if we run into a scenario where the rate limit remaining is 0.01 but I have run the test a couple of times and it has passed each time.
> 
> In fact, before I made that change, I ran the failed test a second time, and it succeeded.

bbengfort commented on pkg/quarterdeck/middleware/ratelimiter.go:
```diff
@@ -0,0 +1,123 @@
+*/
+type IPRateLimiter struct {
+	ips   map[string]*ipInfo //contains a map of IP address to information about its request pattern
+	mu    *sync.RWMutex
```
Rather than make this a `*sync.RWMutex` I'm going to embed it into the struct as a value; this makes it so the mutex can't be copied, which is the source of some subtle bugs.

bbengfort commented on pkg/quarterdeck/middleware/ratelimiter.go:
```diff
@@ -0,0 +1,123 @@
+		// limiters for requests coming from the same IP address due to different port values
+		lim := limiter.GetLimiter(c.ClientIP())
+		if !lim.Allow() {
+			c.Writer.Header().Add("X-RateLimit-Limit", fmt.Sprintf("%v", lim.Burst()))
+			c.Writer.Header().Add("X-RateLimit-Remaining", fmt.Sprintf("%.2f", lim.Tokens()))
+			c.Writer.Header().Add("X-RateLimit-Reset", "1")
```
I'm going to shorten this section up a little bit to use `c.Header` directly and to only call the headers once.

bbengfort commented on pkg/quarterdeck/middleware/ratelimiter_test.go:
```diff
@@ -0,0 +1,89 @@
+			// the third request will be allowed but rate limit remaining will be equal to zero
+		} else if i == 2 {
+			require.Equal(t, http.StatusOK, rep.StatusCode)
+			// beyond the third request all requests will be rejected
```
I'll add the less than or equal check back here; I think it's wise to do that.

bbengfort commented on pkg/quarterdeck/quarterdeck_test.go:
```diff
@@ -77,6 +77,11 @@ func (s *quarterdeckTestSuite) SetupSuite() {
 			RefreshDuration: 20 * time.Minute,
 			RefreshOverlap:  -10 * time.Minute,
 		},
+		RateLimit: config.RateLimitConfig{
+			PerSecond: 20.00,
+			Burst:     120,
+			TTL:       5 * time.Minute,
+		},
```
Good catch!

bbengfort reviewed (approved):
Looks great; nicely done @pdamodaran -- thanks for dealing with those concurrency issues and implementing the double checked locking. I think the cleanup routine is also very good. 

I'm going to add some configuration documentation for devops purposes to this PR and make some minor changes that I commented on below; then I'll go ahead and merge this PR!

bbengfort commented:
On more thought I've also changed the per second default to 10 and the burst to 30; I realized that it was per second and when I suggested 20, I was thinking per minute I think. 10 requests per second is 100 ms per request and we expect about 200ms latencies from the front-end including processing. Although I expect Quarterdeck to respond < 75ms; if the requests are coming in faster than 100ms it means someone is scraping and not coming from the front-end. 

One concern that I had was that the Tenant IP address might be rate-limited. We need to ensure that the `X-Forwarded-For` IP address is set and if not, then set by Tenant; otherwise all requests will be rate limited by a single IP. I'll create a story for this.
